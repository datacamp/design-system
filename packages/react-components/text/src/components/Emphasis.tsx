/* eslint-disable global-require */
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { childrenOfType, computeDataAttributes } from '@datacamp/waffles-utils';
import { css } from '@emotion/react';
import PropTypes from 'prop-types';
import React, { ReactElement, ReactNode } from 'react';

import baseStyle from '../baseStyle';

interface EmphasisProps {
  /**
   * The content to display. Can contain a combination of strings, Text
   * components, and Strong Components.
   */
  children: ReactNode;
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
  /**
   * As with all the other waffles components, dataAttributes can be used to set
   * data- html attributes on the element.
   */
  dataAttributes?: { [key: string]: string };
}

const emphasisStyle = css(baseStyle, {
  fontStyle: 'italic',
  fontWeight: tokens.fontWeight.regular.value,
  lineHeight: tokens.size.font[300].value,
});

const Emphasis = ({
  children,
  className,
  dataAttributes,
}: EmphasisProps): ReactElement => {
  const parsedDataAttributes = computeDataAttributes(dataAttributes);

  return (
    <em className={className} css={emphasisStyle} {...parsedDataAttributes}>
      {children}
    </em>
  );
};

export default Emphasis;

const validChildType = PropTypes.oneOfType([
  childrenOfType(require('./Text')),
  childrenOfType(require('./Strong')),
  childrenOfType(require('../alternateComponents/PlainString')),
  PropTypes.string,
  PropTypes.number,
]);

Emphasis.propTypes = {
  children: PropTypes.oneOfType([
    validChildType,
    PropTypes.arrayOf(validChildType),
  ]),
};
