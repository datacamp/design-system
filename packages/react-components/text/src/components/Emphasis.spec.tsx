import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import React from 'react';

import Emphasis from './Emphasis';

const testText = 'Some example text.';
const testClassName = 'example-class';
const testDataAttributes = { example: 'testdata', item: 'othertest' };

describe('<Emphasis />', () => {
  it('renders', async () => {
    const { container } = await axeRender(
      <Emphasis className={testClassName} dataAttributes={testDataAttributes}>
        {testText}
      </Emphasis>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('sets the text in the span', async () => {
    const { getByText } = await axeRender(<Emphasis>{testText}</Emphasis>);
    expect(getByText(testText)).toBeDefined();
  });

  it('sets the className', async () => {
    const { getByText } = await axeRender(
      <Emphasis className={testClassName}>{testText}</Emphasis>,
    );
    expect(getByText(testText)).toHaveClass(testClassName);
  });

  it('sets the data attributes', async () => {
    const { getByText } = await axeRender(
      <Emphasis dataAttributes={testDataAttributes}>{testText}</Emphasis>,
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-item',
      testDataAttributes.item,
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-example',
      testDataAttributes.example,
    );
  });
});
