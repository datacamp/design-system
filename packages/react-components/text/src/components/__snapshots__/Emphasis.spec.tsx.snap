// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Emphasis /> renders 1`] = `
.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5, 25, 45);
  font-family: Studio-Feixen-Sans,Arial;
  font-style: normal;
  font-style: italic;
  font-weight: 400;
  line-height: 16px;
}

<em
  class="example-class emotion-0"
  data-example="testdata"
  data-item="othertest"
>
  Some example text.
</em>
`;
